databuilder/models/table_lineage.py

```diff
@@ -0,0 +1,104 @@
+import re
+from typing import Any, Dict, List, Union  # noqa: F401
+
+from databuilder.models.neo4j_csv_serde import Neo4jCsvSerializable, \
+    RELATION_START_KEY, RELATION_START_LABEL, RELATION_END_KEY, \
+    RELATION_END_LABEL, RELATION_TYPE, RELATION_REVERSE_TYPE
+
+from databuilder.models.table_metadata import TableMetadata
+
+
+class TableLineage(Neo4jCsvSerializable):
+    # type: (...) -> None
+    """
+    Table Lineage Model. It won't create nodes but create upstream/downstream rels.
+    """
+    LABEL = 'Lineage'
+    KEY_FORMAT = '{db}://{cluster}.{schema}/{tbl}/'
+    ORIGIN_DEPENDENCY_RELATION_TYPE = 'UPSTREAM'
+    DEPENDENCY_ORIGIN_RELATION_TYPE = 'DOWNSTREAM'
+
+    def __init__(self,
+                 db_name,  # type: str
+                 schema_name,  # type: str
+                 table_name,  # type: str
+                 cluster,  # type: str
+                 downstream_deps=None,  # type: List
+                 ):
+        # type: (...) -> None
+        self.db = db_name.lower()
+        self.schema = schema_name.lower()
+        self.table = table_name.lower()
+
+        self.cluster = cluster.lower() if cluster else 'gold'
+        # a list of downstream dependencies, each of which will follow
+        # the same key
+        self.downstream_deps = downstream_deps
+        self._node_iter = iter(self.create_nodes())
+        self._relation_iter = iter(self.create_relation())
+
+    def create_next_node(self):
+        # type: (...) -> Union[Dict[str, Any], None]
+        # return the string representation of the data
+        try:
+            return next(self._node_iter)
+        except StopIteration:
+            return None
+
+    def create_next_relation(self):
+        # type: (...) -> Union[Dict[str, Any], None]
+        try:
+            return next(self._relation_iter)
+        except StopIteration:
+            return None
+
+    def get_table_model_key(self, db, cluster, schema, table):
+        # type: (...) -> str
+        return '{db}://{cluster}.{schema}/{table}'.format(db=db,
+                                                          cluster=cluster,
+                                                          schema=schema,
+                                                          table=table)
+
+    def create_nodes(self):
+        # type: () -> List[Union[Dict[str, Any], None]]
+        """
+        It won't create any node for this model
+        :return:
+        """
+        return []
+
+    def create_relation(self):
+        # type: () -> List[Dict[str, Any]]
+        """
+        Create a list of relation between source table and all the downstream tables
+        :return:
+        """
+        results = []
+        for downstream_tab in self.downstream_deps:
+            # every deps should follow '{db}://{cluster}.{schema}/{table}'
+            # todo: if we change the table uri, we should change here.
+            m = re.match('(\w+)://(\w+)\.(\w+)\/(\w+)', downstream_tab)
```

Comment:
I know REGEX would work, but our key is uri after all. Do you think we can use `urlparsing`?
https://docs.python.org/3/library/urllib.parse.html#url-parsing

Reply:
hey @jinhyukchang , played around with the lib:
```In [1]: from urllib.parse import urlparse

In [2]: a='hive://gold.test_s/test_t'

In [3]: o=urlparse(a)

In [4]: o
Out[4]: ParseResult(scheme='hive', netloc='gold.test_s', path='/test_t', params='', query='', fragment='')

In [5]: o=urlparse('hive.test_s.test_t')

In [6]: o
Out[6]: ParseResult(scheme='', netloc='', path='hive.test_s.test_t', params='', query='', fragment='')```

Try to understand the value of using urllib as I see two issues:
1. we will still parse the path to get cluster name and schema name even with urlparse.
2. urlparse's interface is different for py2 and py3.

So I would like to understand more with this lib. Is it due to the string could be encoded in other env?

Reply:
> hey @jinhyukchang , played around with the lib:
> 
> ```
> 
> In [2]: a='hive://gold.test_s/test_t'
> 
> In [3]: o=urlparse(a)
> 
> In [4]: o
> Out[4]: ParseResult(scheme='hive', netloc='gold.test_s', path='/test_t', params='', query='', fragment='')
> 
> In [5]: o=urlparse('hive.test_s.test_t')
> 
> In [6]: o
> Out[6]: ParseResult(scheme='', netloc='', path='hive.test_s.test_t', params='', query='', fragment='')
> ```
> 
> Try to understand the value of using urllib as I see two issues:
> 
> 1. we will still parse the path to get cluster name and schema name even with urlparse.
> 2. urlparse's interface is different for py2 and py3.
> 
> So I would like to understand more with this lib. Is it due to the string could be encoded in other env?

Yep, it still needs some operation after urlparse. It looks cleaner to me but I am fine with REGEX as well. Will leave it to you!